### PR TITLE
Scale up early when deployment creates new RS

### PR DESCRIPTION
**Pull request by janetkuo**
To distinguish between rollout and scale event, scale up when creating new RS. 

cc @bgrant0607 @nikhiljindal @ironcladlou @kargakis @kubernetes/sig-config

### CODE REVIEW

**Comment by bgrant0607** [comment]
See https://github.com/kubernetes/kubernetes/pull/20273#issuecomment-179681716

**Comment by bgrant0607** [comment]
cc @mqliang

File: pkg/util/deployment/deployment.go [hunk 1]
```diff
@@ -217,3 +219,60 @@ func Revision(rc *api.ReplicationController) (int64, error) {
...
+			return newRC.Spec.Replicas, nil
+		}
+		// Scale up.
+		scaleUpCount := maxTotalPods - currentPodCount
+		// Do not exceed the number of desired replicas.
+		// TODO: Use the min function in pkg/util/integers once it's implemented
+		scaleUpCount = int(math.Min(float64(scaleUpCount), float64(deployment.Spec.Replicas-newRC.Spec.Replicas)))
+		return newRC.Spec.Replicas + scaleUpCount, nil
+	case extensions.RecreateDeploymentStrategyType:
+		return 0, nil
```
Comment (kargakis) [inline pkg/util/deployment/deployment.go]: Should be deployment.Spec.Replicas and not zero for recreate.

File: pkg/util/deployment/deployment.go [hunk 2]
```diff
@@ -217,3 +219,60 @@ func Revision(rc *api.ReplicationController) (int64, error) {
...
+}
+
+// maxSurgeReplicas returns the maximum number of pods that can be scheduled above the original number of
+// pods from calculating deployment's maxSurge. Note that only deployments with RollingUpdate strategy
+// contains maxSurge.
+func maxSurgeReplicas(deployment *extensions.Deployment) (int, error) {
+	if !IsRollingUpdate(deployment) {
+		return 0, fmt.Errorf("Expect strategy type: %v, got %v", extensions.RollingUpdateDeploymentStrategyType, deployment.Spec.Strategy.Type)
+	}
+	if deployment.Spec.Strategy.RollingUpdate == nil {
```
Comment (kargakis) [inline pkg/util/deployment/deployment.go]: this should never be nil
https://github.com/kubernetes/kubernetes/blob/06060d4d77a624a3776f72981643dee2093cdca5/pkg/apis/extensions/v1beta1/defaults.go#L88

Comment (janetkuo) [inline pkg/util/deployment/deployment.go]: Removed

File: pkg/util/deployment/deployment.go (hunk 1 above)
Comment (janetkuo) [inline pkg/util/deployment/deployment.go]: Fixed

**Comment by janetkuo** [comment]
@bgrant0607 PTAL

File: pkg/controller/deployment/deployment_controller.go [hunk 3]
```diff
@@ -718,14 +718,22 @@ func (dc *DeploymentController) getNewRC(deployment extensions.Deployment, maxOl
...
 		dc.rcExpectations.DeleteExpectations(dKey)
 		return nil, fmt.Errorf("error creating replication controller: %v", err)
 	}
+	if newReplicasCount > 0 {
+		dc.eventRecorder.Eventf(&deployment, api.EventTypeNormal, "ScalingRC", "Scaled %s rc %s to %d", "up", createdRC.Name, newReplicasCount)
+	}
 
-	err = dc.updateDeploymentRevision(deployment, newRevision)
-	return createdRC, err
+	return createdRC, dc.updateDeploymentRevision(deployment, newRevision)
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: If this update fails, where will it be retried?

File: pkg/util/deployment/deployment.go [hunk 4]
```diff
@@ -217,3 +219,57 @@ func Revision(rc *api.ReplicationController) (int64, error) {
...
+func IsRollingUpdate(deployment *extensions.Deployment) bool {
+	return deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType
+}
+
+// maxSurgeReplicas returns the maximum number of pods that can be scheduled above the original number of
+// pods from calculating deployment's maxSurge. Note that only deployments with RollingUpdate strategy
+// contains maxSurge.
+func maxSurgeReplicas(deployment *extensions.Deployment) (int, error) {
+	if !IsRollingUpdate(deployment) {
+		return 0, fmt.Errorf("Expect strategy type: %v, got %v", extensions.RollingUpdateDeploymentStrategyType, deployment.Spec.Strategy.Type)
```
Comment (kargakis) [inline pkg/util/deployment/deployment.go]: errors should be lowercased

**Comment by bgrant0607** [comment]
Will need to be moved to ReplicaSet, but I'll take a look first.

**Comment by ghodss** [comment]
Has deployment moved to ReplicaSet in 1.2?

**Comment by smarterclayton** [comment]
Yes

File: pkg/util/deployment/deployment.go [hunk 5]
```diff
@@ -217,3 +219,57 @@ func Revision(rc *api.ReplicationController) (int64, error) {
...
+	}
+	maxSurge, isPercent, err := util.GetIntOrPercentValue(&deployment.Spec.Strategy.RollingUpdate.MaxSurge)
```
Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: Please factor this into a function. We do this in lots of places. util.GetValueFromIntOrPercent(string, int) or something similar.

**Comment by bgrant0607** [comment]
This is just refactoring at this point, not setting the replicas count for a newly created ReplicaSet?

**Comment by bgrant0607** [comment]
Ah, nevermind. It's more clear when expanding the whole file.

File: pkg/controller/deployment/deployment_controller.go [hunk 6]
```diff
@@ -959,6 +957,10 @@ func (dc *DeploymentController) calculateStatus(allRCs []*api.ReplicationControl
 }
 
 func (dc *DeploymentController) scaleRCAndRecordEvent(rc *api.ReplicationController, newScale int, deployment extensions.Deployment) (*api.ReplicationController, error) {
+	// No need to scale
+	if rc.Spec.Replicas == newScale {
+		return rc, nil
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Some callers (eg. line 805, above) want to know if scaling occurred. They would determine that by comparing replicas from before and after?

**Comment by bgrant0607** [comment]
Looks pretty good. A couple comments and needs to be ported to ReplicaSet.

File: pkg/util/deployment/deployment.go (hunk 5 above)
Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: This is being done in another PR, so nevermind.

File: pkg/controller/deployment/deployment_controller.go (hunk 3 above)
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Deployment revision is synced periodically right after getting new RS. 
https://github.com/janetkuo/kubernetes/blob/deployment-new-rc-replicas/pkg/controller/deployment/deployment_controller.go#L599-L605
```go
	// Sync deployment's revision number with new replica set
	if newRS.Annotations != nil && len(newRS.Annotations[deploymentutil.RevisionAnnotation]) > 0 &&
		(deployment.Annotations == nil || deployment.Annotations[deploymentutil.RevisionAnnotation] != newRS.Annotations[deploymentutil.RevisionAnnotation]) {
		if err = dc.updateDeploymentRevision(deployment, newRS.Annotations[deploymentutil.RevisionAnnotation]); err != nil {
			glog.V(4).Infof("Error: %v. Unable to update deployment revision, will retry later.", err)
		}
	}
```

File: pkg/util/deployment/deployment.go (hunk 4 above)
Comment (janetkuo) [inline pkg/util/deployment/deployment.go]: Fixed

File: pkg/controller/deployment/deployment_controller.go (hunk 6 above)
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Made `scaleRCAndRecordEvent` return a `bool` for whether the scale event happen or not.

**Comment by janetkuo** [comment]
All comments addressed and ported to replica set. @bgrant0607 @kargakis PTAL.

File: pkg/util/deployment/deployment.go [hunk 7]
```diff
@@ -233,3 +235,57 @@ func Revision(rs *extensions.ReplicaSet) (int64, error) {
...
+}
+
+// maxSurgeReplicas returns the maximum number of pods that can be scheduled above the original number of
+// pods from calculating deployment's maxSurge. Note that only deployments with RollingUpdate strategy
+// contains maxSurge.
+func maxSurgeReplicas(deployment *extensions.Deployment) (int, error) {
+	if !IsRollingUpdate(deployment) {
+		return 0, fmt.Errorf("expect strategy type: %v, got %v", extensions.RollingUpdateDeploymentStrategyType, deployment.Spec.Strategy.Type)
+	}
+	maxSurge, isPercent, err := util.GetIntOrPercentValue(&deployment.Spec.Strategy.RollingUpdate.MaxSurge)
```
Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: FYI, once #20868 merges, we should update this.

Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: Actually, I meant #21044 here

File: test/e2e/deployment.go [hunk 8]
```diff
@@ -389,8 +389,8 @@ func testRecreateDeployment(f *Framework) {
 	newRS, err := deploymentutil.GetNewReplicaSet(*deployment, c)
 	Expect(err).NotTo(HaveOccurred())
 	Expect(newRS).NotTo(Equal(nil))
-	Expect(events.Items[0].Message).Should(Equal(fmt.Sprintf("Scaled down replica set %s to 0", rsName)))
-	Expect(events.Items[1].Message).Should(Equal(fmt.Sprintf("Scaled up replica set %s to 3", newRS.Name)))
+	Expect(events.Items[0].Message).Should(Equal(fmt.Sprintf("Scaled up replica set %s to 3", newRS.Name)))
+	Expect(events.Items[1].Message).Should(Equal(fmt.Sprintf("Scaled down replica set %s to 0", rsName)))
```
Comment (janetkuo) [inline test/e2e/deployment.go]: Scale up event happens before scale down when Recreate because we set rs.Spec.Replicas = deployment.Spec.Replicas when creating the new RS. 

However, in types.go, it says: 
```go
     // Kill all existing pods before creating new ones.                                                                                                   
     RecreateDeploymentStrategyType DeploymentStrategyType = "Recreate"
```

Should we set it to `0` instead when Recreate?

File: pkg/util/deployment/deployment.go [hunk 9]
```diff
@@ -233,3 +235,57 @@ func Revision(rs *extensions.ReplicaSet) (int64, error) {
...
+		maxTotalPods := deployment.Spec.Replicas + maxSurge
+		if currentPodCount >= maxTotalPods {
+			// Cannot scale up.
+			return newRS.Spec.Replicas, nil
+		}
+		// Scale up.
+		scaleUpCount := maxTotalPods - currentPodCount
+		// Do not exceed the number of desired replicas.
+		// TODO: Use the min function in pkg/util/integers once it's implemented
+		scaleUpCount = int(math.Min(float64(scaleUpCount), float64(deployment.Spec.Replicas-newRS.Spec.Replicas)))
```
Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: Ref #20868

**Comment by bgrant0607** [comment]
LGTM. Please squash and apply the label.

File: test/e2e/deployment.go (hunk 8 above)
Comment (bgrant0607) [inline test/e2e/deployment.go]: Do we wait for the number of pods observed to go to 0 before creating the new ReplicaSet in the case of recreate?

Comment (janetkuo) [inline test/e2e/deployment.go]: @bgrant0607 No. Actually, instead of setting it to `0`, we should scale down first, then get/create new RS (with size `deployment.Spec.Replicas`), and then scale up, during Recreate.

Comment (janetkuo) [inline test/e2e/deployment.go]: *Current implementation is get/create new RS first, and then scale down, and then scale up.

Comment (bgrant0607) [inline test/e2e/deployment.go]: @janetkuo SGTM. Do you want to fix that in this PR, or another PR?

Comment (janetkuo) [inline test/e2e/deployment.go]: I'll fix it in this PR.

**Comment by janetkuo** [comment]
@bgrant0607 PTAL the 2nd commit for syncRecreateDeployment.

File: pkg/util/deployment/deployment.go (hunk 7 above)
Comment (janetkuo) [inline pkg/util/deployment/deployment.go]: Will do.

File: pkg/util/deployment/deployment.go (hunk 9 above)
Comment (janetkuo) [inline pkg/util/deployment/deployment.go]: Will update this.

**Comment by janetkuo** [comment]
@k8s-bot e2e test this issue: #IGNORE (e2e tests aren't started)

File: pkg/controller/deployment/deployment_controller.go [hunk 10]
```diff
@@ -509,6 +510,15 @@ func (dc *DeploymentController) syncRecreateDeployment(deployment extensions.Dep
 		return dc.updateDeploymentStatus(allRSs, newRS, deployment)
 	}
 
+	// If we need to create a new RS, create it now
+	if newRS == nil {
+		newRS, oldRSs, err = dc.getNewAndOldReplicaSets(deployment, true)
```
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: This means we are going to relist all rcs just to create the new. Can you add a TODO to fix this?

Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: I think you meant all RSs. Added TODO.

**Comment by janetkuo** [comment]
@bgrant0607 Rebased and PTAL

File: pkg/util/deployment/deployment.go (hunk 7 above)
Comment (janetkuo) [inline pkg/util/deployment/deployment.go]: #21044 is now merged and this is now updated.

File: pkg/controller/deployment/deployment_controller.go [hunk 11]
```diff
@@ -493,7 +493,8 @@ func (dc *DeploymentController) updateDeploymentAndClearRollbackTo(deployment *e
 }
 
 func (dc *DeploymentController) syncRecreateDeployment(deployment extensions.Deployment) error {
-	newRS, oldRSs, err := dc.getAllReplicaSets(deployment)
+	// Don't create a new RS if not already existed, so that we avoid scaling up before scaling down
+	newRS, oldRSs, err := dc.getAllReplicaSets(deployment, false)
```
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Wouldn't it be simpler if in the place where we create the newRC, check if it's recreate, if yes, make sure spec.replicas is zero instead of passing around booleans?

Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Yes it'd be simpler if we set recreate deployment's RS replicas to 0 when creating it https://github.com/kubernetes/kubernetes/pull/20696/files#diff-a5ab8c704a7cc7eb6ea0616cc18829c5R268, which is how I implemented it earlier. 

But aren't we trying to assign correct number of replicas to new RSs at creation time?

Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: duh, and I asked you to change it from zero which is the right starting value for recreate, right?:)

Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Yes!

**Comment by bgrant0607** [comment]
LGTM.

We can improve it more later if necessary.
